enumerables.rb

```diff
@@ -0,0 +1,273 @@
+module Enumerable
+  def my_each(&block)
+    return to_enum unless block_given?
+
+    i = 0
+    if is_a? Range
+      iterable_array = self.to_a
+    elsif is_a? Array
+      iterable_array = self
+    elsif is_a? Hash
+      iterable_array = self.to_a
+      if block.arity == 0
+        yield
+      elsif block.arity == 1
+        i = 0
+        while i < iterable_array.length
+          yield(iterable_array[i])
+          i += 1
+        end
+        return self
+      elsif block.arity >= 2
+        i = 0
+        while i < iterable_array.length
+          yield(iterable_array[i][0], iterable_array[i][1])
+          i += 1
+        end
+      end
+      self
+    end
+
+    while i <= iterable_array.length - 1
+      yield(iterable_array[i])
+      i += 1
+    end
+    self
+  end
+
+  def my_each_with_index
+    return to_enum unless block_given?
+
+    i = 0
+    while i < length
+      yield(self[i], i)
+      i += 1
+    end
+    self
+  end
+
+  def my_select
+    return to_enum unless block_given?
+
+    object_to_return = []
+    if is_a? Range
+      to_a.my_each { |elt| object_to_return.push(elt) if yield(elt) }
+      return object_to_return
+    end
+    my_each do |element|
+      object_to_return << element if yield(element)
+    end
+    object_to_return
+  end
+
+  def my_all?(*arg, &block)
+    if arg.length == 1
+      i = 0
+      if arg[0].is_a? Class
+        while i < length
+          return false unless self[i].is_a? arg[0]
+
+          i += 1
+        end
+        return true
+      else
+        while i < length
+          return false if self[i] != arg
+
+          i += 1
+        end
+      end
+      return true
+    end
+    block = lambda { |obj| obj } unless block_given?
+
+    i = 0
+    while i < length
+      return false unless block.call(self[i])
+
+      i += 1
+    end
+    true
+  end
```

Comment:
## #my_all?:

```ruby

HIGHEST_VALUE = 9
range = Range.new(5, 50)
false_block = proc { |num| num > HIGHEST_VALUE }
words = %w[dog door rod blade]
array = [2, 34, 43, 54, 68, 93, 3, 14, 62, 51, 28, 38, 34]

```

- [ ] returns false if the block returns false or nil::range

```ruby

  range.my_all?(&false_block) === range.all?(&false_block) => false

```

- [ ] when a Regex is passed as an argument returns true if all of the collection matches the Regex

```ruby

  words.my_all?(/d/) === words.all?(/d/) => false

```

- [ ] when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern

```ruby

  array.my_all?(3) === array.all?(3) => false

```